File: app/routes/admin_routes.py
from flask import Blueprint, request, jsonify
from app.middleware.auth import require_auth
from app.services.db_access import (
    get_all_users, update_user, delete_user,
    get_all_activos, create_activo, update_activo, delete_activo,
    get_all_amenazas, create_amenaza, update_amenaza, delete_amenaza,
    get_all_vulnerabilidades, create_vulnerabilidad, update_vulnerabilidad, delete_vulnerabilidad,
    get_all_salvaguardas, create_salvaguarda, update_salvaguarda, delete_salvaguarda,
    get_all_relaciones, create_relacion, update_relacion, delete_relacion
)

admin_bp = Blueprint('admin', __name__)

# Rutas de usuarios
@admin_bp.route('/users', methods=['GET'])
@require_auth
def get_users():
    if request.user.get('role') != 'admin':
        return jsonify({'error': 'No autorizado'}), 403
    return jsonify(get_all_users())

@admin_bp.route('/users/<user_id>', methods=['PUT'])
@require_auth
def update_user_route(user_id):
    if request.user.get('role') != 'admin':
        return jsonify({'error': 'No autorizado'}), 403
    data = request.get_json()
    return jsonify(update_user(user_id, data))

@admin_bp.route('/users/<user_id>', methods=['DELETE'])
@require_auth
def delete_user_route(user_id):
    if request.user.get('role') != 'admin':
        return jsonify({'error': 'No autorizado'}), 403
    delete_user(user_id)
    return jsonify({'message': 'Usuario eliminado'})

# Rutas de Activos
@admin_bp.route('/activos', methods=['GET'])
@require_auth
def get_activos():
    return jsonify(get_all_activos())

@admin_bp.route('/activos', methods=['POST'])
@require_auth
def create_activo_route():
    data = request.get_json()
    return jsonify(create_activo(data)), 201

@admin_bp.route('/activos/<activo_id>', methods=['PUT'])
@require_auth
def update_activo_route(activo_id):
    data = request.get_json()
    return jsonify(update_activo(activo_id, data))

@admin_bp.route('/activos/<activo_id>', methods=['DELETE'])
@require_auth
def delete_activo_route(activo_id):
    delete_activo(activo_id)
    return jsonify({'message': 'Activo eliminado'})

# Rutas de Amenazas
@admin_bp.route('/amenazas', methods=['GET'])
@require_auth
def get_amenazas():
    return jsonify(get_all_amenazas())

@admin_bp.route('/amenazas', methods=['POST'])
@require_auth
def create_amenaza_route():
    data = request.get_json()
    return jsonify(create_amenaza(data)), 201

@admin_bp.route('/amenazas/<amenaza_id>', methods=['PUT'])
@require_auth
def update_amenaza_route(amenaza_id):
    data = request.get_json()
    return jsonify(update_amenaza(amenaza_id, data))

@admin_bp.route('/amenazas/<amenaza_id>', methods=['DELETE'])
@require_auth
def delete_amenaza_route(amenaza_id):
    delete_amenaza(amenaza_id)
    return jsonify({'message': 'Amenaza eliminada'})

# Rutas de Vulnerabilidades
@admin_bp.route('/vulnerabilidades', methods=['GET'])
@require_auth
def get_vulnerabilidades():
    return jsonify(get_all_vulnerabilidades())

@admin_bp.route('/vulnerabilidades', methods=['POST'])
@require_auth
def create_vulnerabilidad_route():
    data = request.get_json()
    return jsonify(create_vulnerabilidad(data)), 201

@admin_bp.route('/vulnerabilidades/<vulnerabilidad_id>', methods=['PUT'])
@require_auth
def update_vulnerabilidad_route(vulnerabilidad_id):
    data = request.get_json()
    return jsonify(update_vulnerabilidad(vulnerabilidad_id, data))

@admin_bp.route('/vulnerabilidades/<vulnerabilidad_id>', methods=['DELETE'])
@require_auth
def delete_vulnerabilidad_route(vulnerabilidad_id):
    delete_vulnerabilidad(vulnerabilidad_id)
    return jsonify({'message': 'Vulnerabilidad eliminada'})

# Rutas de Salvaguardas
@admin_bp.route('/salvaguardas', methods=['GET'])
@require_auth
def get_salvaguardas():
    return jsonify(get_all_salvaguardas())

@admin_bp.route('/salvaguardas', methods=['POST'])
@require_auth
def create_salvaguarda_route():
    data = request.get_json()
    return jsonify(create_salvaguarda(data)), 201

@admin_bp.route('/salvaguardas/<salvaguarda_id>', methods=['PUT'])
@require_auth
def update_salvaguarda_route(salvaguarda_id):
    data = request.get_json()
    return jsonify(update_salvaguarda(salvaguarda_id, data))

@admin_bp.route('/salvaguardas/<salvaguarda_id>', methods=['DELETE'])
@require_auth
def delete_salvaguarda_route(salvaguarda_id):
    delete_salvaguarda(salvaguarda_id)
    return jsonify({'message': 'Salvaguarda eliminada'})

# Rutas de Relaciones
@admin_bp.route('/relaciones', methods=['GET'])
@require_auth
def get_relaciones():
    return jsonify(get_all_relaciones())

@admin_bp.route('/relaciones', methods=['POST'])
@require_auth
def create_relacion_route():
    data = request.get_json()
    return jsonify(create_relacion(data)), 201

@admin_bp.route('/relaciones/<relacion_id>', methods=['PUT'])
@require_auth
def update_relacion_route(relacion_id):
    data = request.get_json()
    return jsonify(update_relacion(relacion_id, data))

@admin_bp.route('/relaciones/<relacion_id>', methods=['DELETE'])
@require_auth
def delete_relacion_route(relacion_id):
    delete_relacion(relacion_id)
    return jsonify({'message': 'Relación eliminada'}) 

File: src/routes/adminRoutes.js
import express from 'express';
import { verifyToken, isAdmin } from '../middleware/auth.js';
import * as activosService from '../services/activosService.js';
import * as amenazasService from '../services/amenazasService.js';
import * as vulnerabilidadesService from '../services/vulnerabilidadesService.js';
import * as salvaguardasService from '../services/salvaguardasService.js';

const router = express.Router();

// Middleware para verificar token y rol de admin
router.use(verifyToken, isAdmin);

// Rutas para Activos
router.get('/activos', async (req, res) => {
  try {
    const activos = await activosService.getAllActivos();
    res.json(activos);
  } catch (error) {
    res.status(500).json({ error: 'Error al obtener activos' });
  }
});

router.get('/activos/:id', async (req, res) => {
  try {
    const activo = await activosService.getActivoById(req.params.id);
    if (!activo) return res.status(404).json({ error: 'Activo no encontrado' });
    res.json(activo);
  } catch (error) {
    res.status(500).json({ error: 'Error al obtener activo' });
  }
});

router.post('/activos', async (req, res) => {
  try {
    const id = await activosService.createActivo(req.body);
    res.status(201).json({ id });
  } catch (error) {
    res.status(500).json({ error: 'Error al crear activo' });
  }
});

router.put('/activos/:id', async (req, res) => {
  try {
    const activo = await activosService.updateActivo(req.params.id, req.body);
    res.json(activo);
  } catch (error) {
    res.status(500).json({ error: 'Error al actualizar activo' });
  }
});

router.delete('/activos/:id', async (req, res) => {
  try {
    await activosService.removeActivo(req.params.id);
    res.json({ message: 'Activo eliminado correctamente' });
  } catch (error) {
    res.status(500).json({ error: 'Error al eliminar activo' });
  }
});

// Rutas para Amenazas
router.get('/amenazas', async (req, res) => {
  try {
    const amenazas = await amenazasService.getAllAmenazas();
    res.json(amenazas);
  } catch (error) {
    res.status(500).json({ error: 'Error al obtener amenazas' });
  }
});

router.get('/amenazas/:id', async (req, res) => {
  try {
    const amenaza = await amenazasService.getAmenazaById(req.params.id);
    if (!amenaza) return res.status(404).json({ error: 'Amenaza no encontrada' });
    res.json(amenaza);
  } catch (error) {
    res.status(500).json({ error: 'Error al obtener amenaza' });
  }
});

router.post('/amenazas', async (req, res) => {
  try {
    const id = await amenazasService.createAmenaza(req.body);
    res.status(201).json({ id });
  } catch (error) {
    res.status(500).json({ error: 'Error al crear amenaza' });
  }
});

router.put('/amenazas/:id', async (req, res) => {
  try {
    const amenaza = await amenazasService.updateAmenaza(req.params.id, req.body);
    res.json(amenaza);
  } catch (error) {
    res.status(500).json({ error: 'Error al actualizar amenaza' });
  }
});

router.delete('/amenazas/:id', async (req, res) => {
  try {
    await amenazasService.removeAmenaza(req.params.id);
    res.json({ message: 'Amenaza eliminada correctamente' });
  } catch (error) {
    res.status(500).json({ error: 'Error al eliminar amenaza' });
  }
});

// Rutas para Vulnerabilidades
router.get('/vulnerabilidades', async (req, res) => {
  try {
    const vulnerabilidades = await vulnerabilidadesService.getAllVulnerabilidades();
    res.json(vulnerabilidades);
  } catch (error) {
    res.status(500).json({ error: 'Error al obtener vulnerabilidades' });
  }
});

router.get('/vulnerabilidades/:id', async (req, res) => {
  try {
    const vulnerabilidad = await vulnerabilidadesService.getVulnerabilidadById(req.params.id);
    if (!vulnerabilidad) return res.status(404).json({ error: 'Vulnerabilidad no encontrada' });
    res.json(vulnerabilidad);
  } catch (error) {
    res.status(500).json({ error: 'Error al obtener vulnerabilidad' });
  }
});

router.post('/vulnerabilidades', async (req, res) => {
  try {
    const id = await vulnerabilidadesService.createVulnerabilidad(req.body);
    res.status(201).json({ id });
  } catch (error) {
    res.status(500).json({ error: 'Error al crear vulnerabilidad' });
  }
});

router.put('/vulnerabilidades/:id', async (req, res) => {
  try {
    const vulnerabilidad = await vulnerabilidadesService.updateVulnerabilidad(req.params.id, req.body);
    res.json(vulnerabilidad);
  } catch (error) {
    res.status(500).json({ error: 'Error al actualizar vulnerabilidad' });
  }
});

router.delete('/vulnerabilidades/:id', async (req, res) => {
  try {
    await vulnerabilidadesService.removeVulnerabilidad(req.params.id);
    res.json({ message: 'Vulnerabilidad eliminada correctamente' });
  } catch (error) {
    res.status(500).json({ error: 'Error al eliminar vulnerabilidad' });
  }
});

// Rutas para Salvaguardas
router.get('/salvaguardas', async (req, res) => {
  try {
    const salvaguardas = await salvaguardasService.getAllSalvaguardas();
    res.json(salvaguardas);
  } catch (error) {
    res.status(500).json({ error: 'Error al obtener salvaguardas' });
  }
});

router.get('/salvaguardas/:id', async (req, res) => {
  try {
    const salvaguarda = await salvaguardasService.getSalvaguardaById(req.params.id);
    if (!salvaguarda) return res.status(404).json({ error: 'Salvaguarda no encontrada' });
    res.json(salvaguarda);
  } catch (error) {
    res.status(500).json({ error: 'Error al obtener salvaguarda' });
  }
});

router.post('/salvaguardas', async (req, res) => {
  try {
    const id = await salvaguardasService.createSalvaguarda(req.body);
    res.status(201).json({ id });
  } catch (error) {
    res.status(500).json({ error: 'Error al crear salvaguarda' });
  }
});

router.put('/salvaguardas/:id', async (req, res) => {
  try {
    const salvaguarda = await salvaguardasService.updateSalvaguarda(req.params.id, req.body);
    res.json(salvaguarda);
  } catch (error) {
    res.status(500).json({ error: 'Error al actualizar salvaguarda' });
  }
});

router.delete('/salvaguardas/:id', async (req, res) => {
  try {
    await salvaguardasService.removeSalvaguarda(req.params.id);
    res.json({ message: 'Salvaguarda eliminada correctamente' });
  } catch (error) {
    res.status(500).json({ error: 'Error al eliminar salvaguarda' });
  }
});

export default router; 